pkg/api/mcapi.go

```diff
@@ -48,89 +50,144 @@ func NewClient() *Client {
+			Password: myclient.Password,
+		})
+		logger.V(4).Info("==============config=================", "IPAddr", config.MCIpAddress, "Username", config.MCUsername)
+		api_client, err = common.Login(myclient.Ctx, config)
+
+		if err == nil && api_client != nil {
+			myclient.apiClient = api_client
+			configuration := api_client.GetConfig()
+			myclient.SessionKey = configuration.DefaultHeader["sessionKey"]
+		}
+		// if we have a new session key, login was successful and we can leave
+		// if not, try the next controller IP
+		if myclient.SessionKey != "" {
+			break
+		}
 	}
-	_, _, err := client.apiClient.DefaultApi.LogoutGet(client.Ctx).Execute()
 	return err
 }
 
-// SessionValid : Determine if a session is valid, if not a login is required
-// Makes the 'show controller-date' API call to validate the session is still valid
-//
-// Deprecated: This function will be made redundant by retry logic in a future update
-func (client *Client) SessionValid(addr, username string) bool {
+// Logout: Called to log out of the storage controller API
+func (client *Client) Logout() error {
 	if client.Ctx == nil {
-		return false
+		client.Ctx = context.Background()
 	}
-	logger := klog.FromContext(client.Ctx)
-
-	// addr may include protocol and ip address or hostname, or only ip address
-	ipaddress := common.GetAddress(addr)
-
-	if client.Addr == ipaddress && client.Username == username {
-		if client.SessionKey == "" {
-			logger.V(2).Info("session invalid", "sessionkey", client.SessionKey)
-			return false
-		}
+	_, _, err := client.apiClient.DefaultApi.LogoutGet(client.Ctx).Execute()
+	if err == nil {
+		client.SessionKey = ""
 	}
-
-	//run an API call to test that the session is still valid
-	_, _, err := client.apiClient.DefaultApi.ShowControllerDateGet(client.Ctx).Execute()
-	return err == nil
+	return err
 }
 
 // InitSystemInfo: Retrieve and store system information for this client
 func (client *Client) InitSystemInfo() error {
 
-	err := AddSystem(client.Addr, client)
+	err := AddSystem(client.CurrentAddr, client)
 	if err != nil {
-		return fmt.Errorf("unable to add system info for ip (%s) ", client.Addr)
+		return fmt.Errorf("unable to add system info for ip (%s) ", client.CurrentAddr)
 	}
 
-	client.Info, err = GetSystem(client.Addr)
+	client.Info, err = GetSystem(client.CurrentAddr)
 	if err == nil {
 		_ = Log(client.Info)
 	}
 
 	return err
 }
 
+type ResponseType interface {
+	GetStatus() []client.StatusResourceInner
+}
+
+// ExecuteWithFailover: Retry wrapper for the Execute functions of the openapi generated client library
+// If the initial request fails, attempts to login on the secondary controller
+func ExecuteWithFailover[R ResponseType](executeFunc func() (R, *http.Response, error), client *Client) (R, *common.ResponseStatus, *http.Response, error) {
+	logger := klog.FromContext(client.Ctx)
+	logger.V(4).Info("Execute with failover...")
+	response, httpRes, err := executeFunc()
+	if err == nil {
+		status := response.GetStatus()
```

Comment:
Can response be nil?

Reply:
It could be nil, but the OpenAPI generated GetStatus methods all have a nil check that covers that case. 

example:

func (o *EnclosuresObject) GetStatus() []StatusResourceInner {
    if o == nil || IsNil(o.Status) {
        var ret []StatusResourceInner
        return ret
    }
    return o.Status